### ISSUE TITLE
Build/ Rebuild project

### ISSUE THREAD

**Issue by isaacabraham** [ionide/ionide-vscode-fsharp#142]
I'd like to add support for build / rebuild project to the .proj file in the current folder, I guess using MSBuild. What's the best way you'd recommend to add this?

**Comment by Krzysztof-Cieslak** [comment]
1. You need to find msbuild ( I guess something like checking usual installation paths [we do it in case of FSI], and maybe providing setting for specifying path to msbuild), it's much easier in case of mono as xbuild is usually in PATH I think. 

2. Invoke msbuild / xbuild with right parameters (Debug vs Release; Build vs Rebuild vs Clean [and maybe vs CustomTarget])

3. Somehow pick good fsproj file (currently open, give user list of all projects in repo and pick one, or something like that). I guess I would go with 2 options - picking from list, and building "default" (relative path to "default" specified in settings so people can override it on workspace base).

4. What about output path? (We may want to add option to specify it [through settings file]... or do nothing, so path specified in fsproj is used [that's how VS does it])

**Comment by isaacabraham** [comment]
I would just want to start with the simplest thing - look at the current file that's open, find the project that's either in the same folder or upwards of it, and do a debug build. In the future the configuration could be parameterised - you can have multiple configurations (not just debug and release), and I don't want to go through that whole hassle for v1 of this. Same for output path - just do the standard build of the project, the same as doing a project build in Visual Studio.

So, questions for moving forward on this: -

1. Where should settings be stored (for e.g. specifying MSBuild path)?
2. Where should this feature live in terms of this repo (more or less)?
3. What's the easiest way to test this out locally? I don't want a repeat of my PR to the Ionide Paket repo where I needed 4 releases to get it right!

**Comment by cloudRoutine** [comment]
it'd be in this one, the msbuild property info you need is extracted via forge. MsBuild has to be in a standard location otherwise everything else will get broken across ms programming tools. You also need to check the tools version listed in the project files, there are going to be multiple versions of msbuild installed on any system so you need to make sure you use a compatible one. Configurations live both in the solution and the project files, so you need to check the solution file to make sure that if you're trying to build multiple projects under the same configuration settings that it's actually viable. 

I think looking at the file that's currently open and trying to infer the project it's a part of is a bad approach. If they're added contextual menu items (it's coming soon if it's not in already) to the api right clicking on the project file in the file tree and adding a "Build Project" action to the context menu is more reasonable. 

The easiest way to test it locally is by running the extension in debug mode with will launch an extension development host with it installed

![](http://i.imgur.com/SQgCm6E.png)

FYI when you're dealing with msbuild, everything is a hassle

**Comment by isaacabraham** [comment]
@cloudRoutine thanks for this. I want to have it supported with keyboard shortcuts ideally rather than context menus from the file explorer - e.g. SHIFT + F6 as per VS - hence my idea re: inferring from the currently open file.

**Comment by cloudRoutine** [comment]
I thought by default the build hotkeys are setup to build the solution under the active configuration, although the build menus do offer the option to build the project the active file is associated with. the file inference can work in Vs because you have the solutions open and configurations settings are properly handled which means if you have a file open for the active configuration then it makes sense to build it, but in vscode you can get into all kinds of situations that violate those prerequisites and make it confusing as to what the proper course of action should be inferred to. 

That's why I think it's a bad approach, for the time being. Once the full context can be established better then I think it'd be a great feature to have. Also just from a "just get something working" perspective it seems like the menu command approach would be the fastest and easiest way to get something implemented that you can run and test, worry about the other context reliant details later.

**Comment by Krzysztof-Cieslak** [comment]
@isaacabraham:

1. Don't mind @cloudRoutine too much, he loves to over-engineer solutions ( ;-) ), I think we can start with some simple usecase, :-)

> Where should settings be stored (for e.g. specifying MSBuild path)?

It should be defined as VSCode settings (example of definition - https://github.com/ionide/ionide-vscode-fsharp/blob/master/release/package.json#L196-L200, example of usage in source code - https://github.com/ionide/ionide-vscode-fsharp/blob/master/src/Components/Environment.fs#L72-L73)

> Where should this feature live in terms of this repo (more or less)?

New file under Components folder, most of them are just module with public activate function where you define commands and/or other stuff. Take a look at Forge.fs for good example (as it only defines commands, uses `showQuickPick` which will be probably used for picking project to build, etc)

> What's the easiest way to test this out locally? I don't want a repeat of my PR to the Ionide Paket repo where I needed 4 releases to get it right!

Open repo (with your changes) in VSCode and press F5 - it will build everything and start experimental VSCode instance.

**Comment by cloudRoutine** [comment]
Specifying the build tools location might be more appropriate for mono (I have no idea what their conventions are), but they should still be automatically detected on windows.

the setting should be an array so it can store multiple values.

```json
        "FSharp.msbuildLocations": {
            "type": "array",
            "default": [],
            "description": "Directories containing MSBuild tools"            
        }
```

@Krzysztof-Cieslak I was actually presenting the simplest use case as the starting point

**Comment by vasily-kirichenko** [comment]
Why just not start `msbuild <project file name the current file belongs to>`? This is #1 of features I want.

**Comment by vasily-kirichenko** [comment]
Another idea: is it possible to define a FAKE `Target` named `Build current project`, which would find the project file the current file belongs to, then run msbuild on it? Looks like a simple solution.

@forki

**Comment by baronfel** [comment]
Ok, I've submitted an initial PR for basic `build current` and `build selected` project tasks.  What else would you all like to see in terms of 
* specific target invocation
* specific msbuild Configuration (debug/release)?

**Comment by Krzysztof-Cieslak** [comment]
For sure we should ask Rebuild and Clean commands so it has same functionality as VS.

Not sure how to handle Configurations (and it's bigger problem than just building [right now we don't pass configuration to ProjectCracker so it's invoked with default settings defined in project files]). I think for now, we can just stay with passing no Configuration (i.e using defaults defined in project file), until there is some more general way of handling MsBuild configurations in Ionide.

**Comment by baronfel** [comment]
Which project would rebuild and clean target? Would they each have current/chosen variants?

**Comment by baronfel** [comment]
Also, any preferences for keybindings?

**Comment by Krzysztof-Cieslak** [comment]
Yeah, I guess we should have current and chosen for each target.

Don't really have preferences for keybindings (it doesn't really matter, it's super easy to change them for users) as long as they don't override some important VSCode commands. ;)

**Comment by baronfel** [comment]
Is there a way to map the output of this panel to the built-in msbuild error reporter?  I know that if you configure a task with that reporter you get errors in the pane for free, but this isn't a task per se

**Comment by Krzysztof-Cieslak** [comment]
I don't think we can do this right now.

**Comment by Krzysztof-Cieslak** [comment]
I know Code Team is working on some changes in how tasks work and adding extension APIs for them, so maybe we will be able to move MsBuild support there in the future but it's not here yet.

**Comment by baronfel** [comment]
ok, rebuild and clean are done.

**Comment by Krzysztof-Cieslak** [comment]
OK, initial support was released in `2.20.0`.
